File: backend/app/services/resource_service.py
import asyncio
import html
import json
import logging
import re
import shutil
import urllib.error
import urllib.request
import uuid
from datetime import UTC, datetime
from pathlib import Path

from fastapi import HTTPException

from app.schemas.resource import Resource, ResourceCreate, ResourceType
from app.services import ai_service
from app.services.note_service import note_exists
from app.services.workspace_service import get_workspace_dir

logger = logging.getLogger("app")

METADATA_FILENAME = "resources.jsonl"
FETCH_TIMEOUT_SECONDS = 10
TAG_PATTERN = re.compile(r"<[^>]+>")
WHITESPACE_PATTERN = re.compile(r"[ \t]+")


def _metadata_path(workspace_dir: Path) -> Path:
    return workspace_dir / METADATA_FILENAME


def _resource_dir(workspace_dir: Path, resource_id: str) -> Path:
    return workspace_dir / "resources" / resource_id


def _read_all(workspace_dir: Path) -> list[dict]:
    path = _metadata_path(workspace_dir)
    if not path.exists():
        return []
    return [json.loads(line) for line in path.read_text().splitlines() if line.strip()]


def _write_all(workspace_dir: Path, entries: list[dict]) -> None:
    _metadata_path(workspace_dir).write_text("".join(json.dumps(entry) + "\n" for entry in entries))


def _find(entries: list[dict], resource_id: str) -> dict:
    # Same reasoning as note_service._find: resource_id is only ever checked
    # against this per-workspace index, which only ever contains ids we
    # generated ourselves, so it can't be used to escape the resources dir.
    for entry in entries:
        if entry["id"] == resource_id:
            return entry
    raise HTTPException(status_code=404, detail="Resource not found")


def _to_resource(entry: dict) -> Resource:
    return Resource(**entry)


def _require_note(workspace_id: str, note_id: str) -> None:
    if not note_exists(workspace_id, note_id):
        raise HTTPException(status_code=404, detail="Note not found")


def list_resources(workspace_id: str, note_id: str) -> list[Resource]:
    _require_note(workspace_id, note_id)
    workspace_dir = get_workspace_dir(workspace_id)
    entries = [e for e in _read_all(workspace_dir) if e["note_id"] == note_id]
    entries.sort(key=lambda e: e["created_at"])
    return [_to_resource(e) for e in entries]


def create_url_resource(workspace_id: str, note_id: str, data: ResourceCreate) -> Resource:
    _require_note(workspace_id, note_id)
    title = data.title.strip()
    if not title:
        raise HTTPException(status_code=422, detail="Resource title cannot be empty")

    workspace_dir = get_workspace_dir(workspace_id)
    resource_id = uuid.uuid4().hex
    now = datetime.now(UTC).isoformat()
    entry = {
        "id": resource_id,
        "note_id": note_id,
        "type": data.type,
        "source": data.source,
        "title": title,
        "status": "queued",
        "error": None,
        "created_at": now,
        "updated_at": now,
    }

    entries = _read_all(workspace_dir)
    entries.append(entry)
    _write_all(workspace_dir, entries)
    _resource_dir(workspace_dir, resource_id).mkdir(parents=True, exist_ok=True)

    return _to_resource(entry)


def create_file_resource(
    workspace_id: str,
    note_id: str,
    resource_type: ResourceType,
    title: str,
    filename: str,
    data: bytes,
) -> Resource:
    _require_note(workspace_id, note_id)
    title = title.strip()
    if not title:
        raise HTTPException(status_code=422, detail="Resource title cannot be empty")

    workspace_dir = get_workspace_dir(workspace_id)
    resource_id = uuid.uuid4().hex
    now = datetime.now(UTC).isoformat()
    entry = {
        "id": resource_id,
        "note_id": note_id,
        "type": resource_type,
        "source": filename,
        "title": title,
        "status": "completed",
        "error": None,
        "created_at": now,
        "updated_at": now,
    }

    entries = _read_all(workspace_dir)
    entries.append(entry)
    _write_all(workspace_dir, entries)

    resource_dir = _resource_dir(workspace_dir, resource_id)
    resource_dir.mkdir(parents=True, exist_ok=True)
    suffix = Path(filename).suffix
    (resource_dir / f"original{suffix}").write_bytes(data)

    return _to_resource(entry)


def delete_resource(workspace_id: str, resource_id: str) -> None:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, resource_id)
    entries.remove(entry)
    _write_all(workspace_dir, entries)
    shutil.rmtree(_resource_dir(workspace_dir, resource_id), ignore_errors=True)


def _set_status(workspace_id: str, resource_id: str, status: str, error: str | None = None) -> None:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, resource_id)
    entry["status"] = status
    entry["error"] = error
    entry["updated_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)


def _fetch_and_extract(url: str) -> str:
    request = urllib.request.Request(url, headers={"User-Agent": "AILearningWorkspace/0.1"})
    with urllib.request.urlopen(request, timeout=FETCH_TIMEOUT_SECONDS) as response:  # noqa: S310
        charset = response.headers.get_content_charset() or "utf-8"
        raw = response.read().decode(charset, errors="replace")
        content_type = response.headers.get_content_type()

    if content_type == "text/html":
        text = re.sub(r"(?is)<(script|style).*?>.*?</\1>", " ", raw)
        text = TAG_PATTERN.sub(" ", text)
        text = html.unescape(text)
        text = WHITESPACE_PATTERN.sub(" ", text)
        text = "\n".join(line.strip() for line in text.splitlines() if line.strip())
        return text

    return raw


async def process_resource(workspace_id: str, resource_id: str) -> None:
    # Reading = fetching the raw response; processing = stripping it to
    # plain text, then (Milestone 12) asking the AI for a faithful summary.
    workspace_dir = get_workspace_dir(workspace_id)
    entry = _find(_read_all(workspace_dir), resource_id)
    url = entry["source"]

    _set_status(workspace_id, resource_id, "reading")
    try:
        _set_status(workspace_id, resource_id, "processing")
        text = await asyncio.to_thread(_fetch_and_extract, url)
    except (urllib.error.URLError, TimeoutError, ValueError) as err:
        _set_status(workspace_id, resource_id, "failed", error=str(err))
        return

    resource_dir = _resource_dir(workspace_dir, resource_id)
    (resource_dir / "content.txt").write_text(text)

    # A summary is a bonus on top of the raw fetched content — if the AI
    # call fails (no API key configured, rate limited, etc.) the resource
    # is still fully usable via its raw content, so this doesn't fail the
    # whole resource.
    try:
        summary = await ai_service.run_action("process-resource", text)
        (resource_dir / "summary.txt").write_text(summary)
    except Exception:
        logger.exception("AI summary failed for resource %s", resource_id)

    _set_status(workspace_id, resource_id, "completed")

File: backend/src/services/resourceService.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { readJsonl, writeJsonl } from '../utils/fileDb.js';
import { getWorkspaceDir } from './workspaceService.js';
import { noteExists } from './noteService.js';
import { runAction } from './aiService.js';
import { Resource, ResourceCreate, ResourceType } from '../types/resource.js';

const METADATA_FILENAME = 'resources.jsonl';
const FETCH_TIMEOUT_MS = 10000;
const RESOURCE_CONTEXT_CHAR_LIMIT = 3000;

function metadataPath(workspaceDir: string): string {
  return path.join(workspaceDir, METADATA_FILENAME);
}

function resourceDir(workspaceDir: string, resourceId: string): string {
  return path.join(workspaceDir, 'resources', resourceId);
}

function readAllResources(workspaceDir: string): any[] {
  return readJsonl(metadataPath(workspaceDir));
}

function writeAllResources(workspaceDir: string, entries: any[]): void {
  writeJsonl(metadataPath(workspaceDir), entries);
}

function findResource(entries: any[], resourceId: string): any {
  const entry = entries.find(e => e.id === resourceId);
  if (!entry) {
    throw { status: 404, message: 'Resource not found' };
  }
  return entry;
}

function requireNote(workspaceId: string, noteId: string): void {
  if (!noteExists(workspaceId, noteId)) {
    throw { status: 404, message: 'Note not found' };
  }
}

export function listResources(workspaceId: string, noteId: string): Resource[] {
  requireNote(workspaceId, noteId);
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllResources(workspaceDir).filter(e => e.note_id === noteId);
  return entries.sort((a, b) => new Date(a.created_at).getTime() - new Date(b.created_at).getTime());
}

export function getResourceContext(workspaceId: string, noteId: string): string {
  const resources = listResources(workspaceId, noteId).filter(r => r.status === 'completed');
  if (resources.length === 0) return '';

  const workspaceDir = getWorkspaceDir(workspaceId);
  const blocks: string[] = [];

  for (const resource of resources) {
    const dir = resourceDir(workspaceDir, resource.id);
    let textPath = path.join(dir, 'summary.txt');
    if (!fs.existsSync(textPath)) {
      textPath = path.join(dir, 'content.txt');
    }
    if (!fs.existsSync(textPath)) continue;

    const text = fs.readFileSync(textPath, 'utf8').trim().slice(0, RESOURCE_CONTEXT_CHAR_LIMIT);
    if (text) {
      blocks.push(`### ${resource.title}\n${text}`);
    }
  }

  return blocks.join('\n\n');
}

export function createUrlResource(workspaceId: string, noteId: string, data: ResourceCreate): Resource {
  requireNote(workspaceId, noteId);
  const title = data.title.trim();
  if (!title) {
    throw { status: 422, message: 'Resource title cannot be empty' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const resourceId = crypto.randomUUID().replace(/-/g, '');
  const now = new Date().toISOString();

  const entry = {
    id: resourceId,
    note_id: noteId,
    type: data.type,
    source: data.source,
    title,
    status: 'queued' as const,
    error: null as string | null,
    created_at: now,
    updated_at: now
  };

  const entries = readAllResources(workspaceDir);
  entries.push(entry);
  writeAllResources(workspaceDir, entries);

  fs.mkdirSync(resourceDir(workspaceDir, resourceId), { recursive: true });

  return entry;
}

export function createFileResource(
  workspaceId: string,
  noteId: string,
  resourceType: ResourceType,
  title: string,
  filename: string,
  fileData: Buffer
): Resource {
  requireNote(workspaceId, noteId);
  const trimmedTitle = title.trim();
  if (!trimmedTitle) {
    throw { status: 422, message: 'Resource title cannot be empty' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const resourceId = crypto.randomUUID().replace(/-/g, '');
  const now = new Date().toISOString();

  const entry = {
    id: resourceId,
    note_id: noteId,
    type: resourceType,
    source: filename,
    title: trimmedTitle,
    status: 'completed' as const,
    error: null as string | null,
    created_at: now,
    updated_at: now
  };

  const entries = readAllResources(workspaceDir);
  entries.push(entry);
  writeAllResources(workspaceDir, entries);

  const dir = resourceDir(workspaceDir, resourceId);
  fs.mkdirSync(dir, { recursive: true });
  const ext = path.extname(filename);
  fs.writeFileSync(path.join(dir, `original${ext}`), fileData);

  return entry;
}

export function getResourceFilePath(workspaceId: string, resourceId: string): string | null {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const dir = resourceDir(workspaceDir, resourceId);
  if (!fs.existsSync(dir)) return null;

  const files = fs.readdirSync(dir);
  const originalFile = files.find(f => f.startsWith('original.'));
  return originalFile ? path.join(dir, originalFile) : null;
}

export function deleteResource(workspaceId: string, resourceId: string): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllResources(workspaceDir);
  const entry = entries.find(e => e.id === resourceId);
  if (entry) {
    const index = entries.indexOf(entry);
    if (index > -1) {
      entries.splice(index, 1);
    }
    writeAllResources(workspaceDir, entries);
  }
  const dir = resourceDir(workspaceDir, resourceId);
  if (fs.existsSync(dir)) {
    fs.rmSync(dir, { recursive: true, force: true });
  }
}

function setStatus(workspaceId: string, resourceId: string, status: string, error: string | null = null): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllResources(workspaceDir);
  const entry = findResource(entries, resourceId);
  entry.status = status;
  entry.error = error;
  entry.updated_at = new Date().toISOString();
  writeAllResources(workspaceDir, entries);
}

function htmlUnescape(str: string): string {
  return str
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'")
    .replace(/&#(\d+);/g, (_, dec) => String.fromCharCode(parseInt(dec, 10)))
    .replace(/&#x([0-9a-f]+);/gi, (_, hex) => String.fromCharCode(parseInt(hex, 16)));
}

async function fetchAndExtract(urlStr: string): Promise<string> {
  const controller = new AbortController();
  const id = setTimeout(() => controller.abort(), FETCH_TIMEOUT_MS);
  
  try {
    const response = await fetch(urlStr, {
      headers: { 'User-Agent': 'AILearningWorkspace/0.1' },
      signal: controller.signal
    });
    clearTimeout(id);

    if (!response.ok) {
      throw new Error(`Fetch failed with status ${response.status}`);
    }

    const contentType = response.headers.get('content-type') || '';
    const buffer = await response.arrayBuffer();
    // Default to utf-8
    const decoder = new TextDecoder('utf-8');
    const raw = decoder.decode(buffer);

    if (contentType.includes('text/html')) {
      let text = raw.replace(/<(script|style)[\s\S]*?>[\s\S]*?<\/\1>/gi, ' ');
      text = text.replace(/<[^>]+>/g, ' ');
      text = htmlUnescape(text);
      text = text.replace(/[ \t]+/g, ' ');
      return text
        .split('\n')
        .map(line => line.trim())
        .filter(line => line.length > 0)
        .join('\n');
    }

    return raw;
  } catch (err: any) {
    clearTimeout(id);
    throw err;
  }
}

export async function processResource(workspaceId: string, resourceId: string): Promise<void> {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entry = findResource(readAllResources(workspaceDir), resourceId);
  const url = entry.source;

  setStatus(workspaceId, resourceId, 'reading');
  let text = '';
  try {
    setStatus(workspaceId, resourceId, 'processing');
    text = await fetchAndExtract(url);
  } catch (err: any) {
    setStatus(workspaceId, resourceId, 'failed', err.message || String(err));
    return;
  }

  const dir = resourceDir(workspaceDir, resourceId);
  fs.writeFileSync(path.join(dir, 'content.txt'), text, 'utf8');

  try {
    const summary = await runAction('process-resource', text);
    fs.writeFileSync(path.join(dir, 'summary.txt'), summary, 'utf8');
  } catch (err) {
    console.error(`[backend] AI summary failed for resource ${resourceId}:`, err);
  }

  setStatus(workspaceId, resourceId, 'completed');
}
